File: flink-scala-code/src/main/scala/com/atguigu/day1/WordCountFromBatch.scala
package com.atguigu.day1

// 导入一些隐式类型转换
import org.apache.flink.streaming.api.scala._

object WordCountFromBatch {

  def main(args: Array[String]): Unit = {
    // 获取运行时环境
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    // 并行度设置为1，所有计算都在同一个分区执行
    env.setParallelism(1)

    val stream = env
      .fromElements(
        "hello word",
        "hello world"
      )
      // 使用空格分割字符串 `\\s`表示空格
      .flatMap(w => w.split("\\s"))
      // 相当于MapReduce中的Map操作
      .map(w => WordCount(w, 1))
      // shuffle操作
      .keyBy(_.word)
      // 聚合`count`字段
      .sum(1)

    stream.print()

    env.execute()
  }

  case class WordCount(word: String, count: Int)
}

File: flink-code-java-0421/src/main/java/com/atguigu/day01/WordCountFromBatch.java
package com.atguigu.day01;

import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.util.Collector;

public class WordCountFromBatch {
    public static void main(String[] args) throws Exception {
        // runtime context
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        // num of task parallelism
        env.setParallelism(1);

        DataStream<String> stream = env.fromElements("hello world", "hello world world");

        SingleOutputStreamOperator<Tuple2<String, Integer>> result = stream
                // Map
                .flatMap(new FlatMapFunction<String, Tuple2<String, Integer>>() {
                    @Override
                    public void flatMap(String s, Collector<Tuple2<String, Integer>> collector) throws Exception {
                        String[] arr = s.split(" ");
                        for (String word : arr) {
                            // new Tuple2<String, Integer>(word, 1) <==> Tuple2.of(word, 1)
                            collector.collect(Tuple2.of(word, 1));
                        }
                    }
                })
                // shuffle
                // r.f0 <==> r._1
                .keyBy(r -> r.f0)
                // reduce
                .sum(1);

        result.print();

        // execute task
        env.execute();
    }
}
